packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +465,106 @@ export default {
     // host-by-host install
     throw new Error('non pool-wide install not implemented')
   },
+
+  async rollingPoolUpdate() {
+    let hosts = filter(this.objects.all, { $type: 'host' })
+    await Promise.all(hosts.map(host => host.$call('assert_can_evacuate')))
+
+    log.debug('Install patches')
+    await this.installPatches()
+
+    // Remember on which hosts the running VMs are
+    const vmsByHost = mapValues(
+      groupBy(
+        filter(this.objects.all, {
+          $type: 'VM',
+          power_state: 'Running',
+          is_control_domain: false,
+        }),
+        vm => {
+          const hostId = vm.$resident_on?.$id
+
+          if (hostId === undefined) {
+            throw new Error('Could not find host of all running VMs')
+          }
+
+          return hostId
+        }
+      ),
+      vms => vms.map(vm => vm.$id)
+    )
+
+    // Cache $fields to make sure we can access them later
+    hosts = hosts.map(host => ({
+      ...host,
+      id: host.$id,
+      ref: host.$ref,
+      metricsId: host.$metrics.$id,
```

Comment:
Why this fields are needed?

Reply:
`waitObjectState` supports `uuids/ids/refs`, so,  you can use `host.uuid` and `host.metrics`.